### PR TITLE
Taken from allies before you

### PR DESCRIPTION
**Pull request by Regisle**
Reworks how taken from allies before you is handled to make it more generic to make it easier to support more sources in the future (like `Vaal Rejuvenation Totem`), as well as adding support for 
`The Jinxed Juju` mod `10% of Damage from Hits is taken from your Spectres' Life before you` 
and the totem mastery `5% of Damage from Hits is taken from your nearest Totem's Life before you`

### CODE REVIEW

File: src/Modules/CalcDefence.lua [hunk 1]
```diff
@@ -89,8 +89,23 @@ function calcs.reducePoolsByDamage(poolTable, damageTable, actor)
 	local output = actor.output
 	local modDB = actor.modDB
 	local poolTbl = poolTable or { }
-	local frostShield = poolTbl.FrostShieldLife or output.FrostShieldLife or 0
-	local soulLink = poolTbl.SoulLink or output.SoulLink or 0
+	
+	local alliesTakenBeforeYou = poolTbl.alliesTakenBeforeYou
+	if not alliesTakenBeforeYou then
+		alliesTakenBeforeYou = {}
+		if output.FrostShieldLife then
```
Comment (QuickStick123) [inline src/Modules/CalcDefence.lua:96]: Should frost shield be grouped inside a allies take damage before you array?

File: src/Modules/CalcDefence.lua [hunk 2]
```diff
@@ -225,8 +238,7 @@ function calcs.reducePoolsByDamage(poolTable, damageTable, actor)
 	end
 
 	return {
-		FrostShieldLife = frostShield,
-		SoulLink = soulLink,
+		alliesTakenBeforeYou = alliesTakenBeforeYou,
```
Comment (QuickStick123) [inline src/Modules/CalcDefence.lua:241]:
```suggestion
		AlliesTakenBeforeYou = alliesTakenBeforeYou,
```
The rest of the fields in this return type are capitalised consider doing the same here.

File: src/Modules/CalcDefence.lua [hunk 3]
```diff
@@ -89,8 +89,23 @@ function calcs.reducePoolsByDamage(poolTable, damageTable, actor)
...
+		if output.FrostShieldLife then
+			alliesTakenBeforeYou["frostShield"] = { remaining = output.FrostShieldLife, percent = output.FrostShieldDamageMitigation / 100 }
+		end
+		if output.TotalSpectreLife then
+			alliesTakenBeforeYou["specters"] = { remaining = output.TotalSpectreLife, percent = output.SpectreAllyDamageMitigation / 100 }
+		end
+		if output.TotalTotemLife then
+			alliesTakenBeforeYou["totems"] = { remaining = output.TotalTotemLife, percent = output.TotemAllyDamageMitigation / 100 }
+		end
+		if output.SoulLink then
```
Comment (QuickStick123) [inline src/Modules/CalcDefence.lua:105]: Consider readding the comment that soul link isn't implemented somewhere.

File: src/Modules/ConfigOptions.lua [hunk 4]
```diff
@@ -492,6 +492,12 @@ Huge sets the radius to 11.
 	{ var = "enemyRadius", type = "integer", label = "Enemy radius:", ifSkill = { "Seismic Trap", "Lightning Spire Trap", "Explosive Trap" }, tooltip = "Configure the radius of an enemy hitbox to calculate some area overlapping (shotgunning) effects.", apply = function(val, modList, enemyModList)
 		modList:NewMod("EnemyRadius", "OVERRIDE", m_max(val, 1), "Config")
 	end },
+	{ var = "TotalSpectreLife", type = "integer", label = "Total Spectre Life:", tooltip = "The total life of your Spectres that can be taken before yours (used by jinxed juju)", apply = function(val, modList, enemyModList)
+		modList:NewMod("TotalSpectreLife", "BASE", val, "Config")
+	end },
+	{ var = "TotalTotemLife", type = "integer", label = "Total Totem Life:", tooltip = "The total life of your Totems (excluding Vaal Rejuvenation Totem) that can be taken before yours (used by totem mastery)", apply = function(val, modList, enemyModList)
+		modList:NewMod("TotalTotemLife", "BASE", val, "Config")
+	end },
```
Comment (QuickStick123) [inline src/Modules/ConfigOptions.lua:500]:
```suggestion
	{ var = "TotalSpectreLife", type = "integer", label = "Total Spectre Life:", ifMod = "takenFromSpectresBeforeYou", ifSkill = "Raise Spectre", tooltip = "The total life of your Spectres that can be taken before yours (used by jinxed juju)", apply = function(val, modList, enemyModList)
		modList:NewMod("TotalSpectreLife", "BASE", val, "Config")
	end },
	{ var = "TotalTotemLife", type = "integer", label = "Total Totem Life:", ifFlag = "Condition:HaveTotem", ifMod = "takenFromTotemsBeforeYou", tooltip = "The total life of your Totems (excluding Vaal Rejuvenation Totem) that can be taken before yours (used by totem mastery)", apply = function(val, modList, enemyModList)
		modList:NewMod("TotalTotemLife", "BASE", val, "Config")
	end },
```

File: src/Modules/ModParser.lua [hunk 5]
```diff
@@ -3679,6 +3679,8 @@ local specialModList = {
 	["(%d+)%% increased armour per second you've been stationary, up to a maximum of (%d+)%%"] = function(num, _, limit) return {
 		mod("Armour", "INC", num, { type = "Multiplier", var = "StationarySeconds", limit = tonumber(limit / num) }, { type = "Condition", var = "Stationary" }),
 	} end,
+	["(%d+)%% of damage from hits is taken from your spectres' life before you"] = function(num) return { mod("takenFromSpectresBeforeYou", "BASE", num) } end,
+	["(%d+)%% of damage from hits is taken from your nearest totem's life before you"] = function(num) return { mod("takenFromTotemsBeforeYou", "BASE", num) } end,
```
Comment (QuickStick123) [inline src/Modules/ModParser.lua:3683]:
```suggestion
	["(%d+)%% of damage from hits is taken from your nearest totem's life before you"] = function(num) return { mod("takenFromTotemsBeforeYou", "BASE", num, { type = "Condition", var = "HaveTotem" }) } end,
```

**Comment by QuickStick123** [comment]
The calculations look correct afaik.

File: src/Modules/ConfigOptions.lua (hunk 4 above)
Comment (deathbeam) [reply src/Modules/ConfigOptions.lua:500]: isnt the ifFlag Condition:HaveTotem same as ifCond HaveTotem?

Comment (QuickStick123) [reply src/Modules/ConfigOptions.lua:500]: ifCond was pulling the condition from the mods and showing it based on that. This makes the totemLife thing toggle based upon that config option.

File: src/Modules/CalcDefence.lua (hunk 1 above)
Comment (Regisle) [reply src/Modules/CalcDefence.lua:96]: frost shield is its own entity and doesnt count as you taking the damage, hence being grouped as a "non-you" entity

File: src/Modules/ConfigOptions.lua (hunk 4 above)
Comment (deathbeam) [reply src/Modules/ConfigOptions.lua:500]: if it should be specifically just based on the config option shouldnt it be ifOption then

Comment (QuickStick123) [reply src/Modules/ConfigOptions.lua:500]: Oh didn't know that was a thing. Might be nice to add these to the docs somewhere with details on what they do.

Comment (QuickStick123) [reply src/Modules/ConfigOptions.lua:500]: Updated my comment with that instead it doesn't seem to force display the config box so still need a condition on the actual mod.

**Comment by Nightblade** [comment]
(added `rejuv` to `pob-dict.txt`)

**Review by LocalIdentity — APPROVED** [review]

**Pull request merged**